xinference/core/image_interface.py

```diff
@@ -0,0 +1,239 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import io
+import logging
+import os
+from typing import Dict, List, Optional, Union
+
+import gradio as gr
+import PIL.Image
+from gradio import Markdown
+
+from ..client.restful.restful_client import RESTfulImageModelHandle
+
+logger = logging.getLogger(__name__)
+
+
+class ImageInterface:
+    def __init__(
+        self,
+        endpoint: str,
+        model_uid: str,
+        model_family: str,
+        model_name: str,
+        model_id: str,
+        model_revision: str,
+        controlnet: Union[None, List[Dict[str, Union[str, None]]]],
+        access_token: Optional[str],
+    ):
+        self.endpoint = endpoint
+        self.model_uid = model_uid
+        self.model_family = model_family
+        self.model_name = model_name
+        self.model_id = model_id
+        self.model_revision = model_revision
+        self.controlnet = controlnet
+        self.access_token = (
+            access_token.replace("Bearer ", "") if access_token is not None else None
+        )
+
+    def build(self) -> gr.Blocks:
+        assert "stable_diffusion" in self.model_family
+
+        interface = self.build_main_interface()
+        interface.queue()
+        # Gradio initiates the queue during a startup event, but since the app has already been
+        # started, that event will not run, so manually invoke the startup events.
+        # See: https://github.com/gradio-app/gradio/issues/5228
+        interface.startup_events()
+        favicon_path = os.path.join(
+            os.path.dirname(os.path.abspath(__file__)),
+            os.path.pardir,
+            "web",
+            "ui",
+            "public",
+            "favicon.svg",
+        )
+        interface.favicon_path = favicon_path
+        return interface
+
+    def text2image_interface(self) -> "gr.Blocks":
+        def text_generate_image(
+            prompt: str,
+            negative_prompt: str,
+            n: int,
+            size_width: int,
+            size_height: int,
+        ) -> PIL.Image.Image:
+            from ..client import RESTfulClient
+
+            client = RESTfulClient(self.endpoint)
+            client._set_token(self.access_token)
+            model = client.get_model(self.model_uid)
+            assert isinstance(model, RESTfulImageModelHandle)
+
+            size = f"{int(size_width)}*{int(size_height)}"
+
+            image_urls = model.text_to_image(
+                prompt=prompt,
+                negative_prompt=negative_prompt,
+                n=n,
+                size=size,
+            )
+
+            logger.info(f"Image URLs: {image_urls}")
+            images = [PIL.Image.open(url["url"]) for url in image_urls["data"]]
```

Comment:
Should use `response_format="b64_json"` to generate images here. 
When xinference is distributed, the `url` must be a file uri on a worker. Supervisor cannot get this uri in this situation. You should pass the base64 string to the frontend，and let the frontend to parse out images.

Reply:
fixed